library/core/src/num/mod.rs

```diff
@@ -1032,29 +1028,52 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     // of multi-byte sequences
     let src = src.as_bytes();
 
-    let (is_positive, digits) = match src[0] {
-        b'+' | b'-' if src[1..].is_empty() => {
-            return Err(PIE { kind: InvalidDigit });
+    let (first, mut digits) = (*src.get(0).ok_or_else(|| PIE { kind: Empty })?, &src[1..]);
+
+    let (is_positive, mut result) = match first {
+        b'+' => {
+            let first = digits.get(0).ok_or_else(|| PIE { kind: InvalidDigit })?;
+            digits = &digits[1..];
+            (
+                true,
+                T::from_u32(
+                    (*first as char).to_digit(radix).ok_or_else(|| PIE { kind: InvalidDigit })?,
+                ),
+            )
         }
-        b'+' => (true, &src[1..]),
-        b'-' if is_signed_ty => (false, &src[1..]),
-        _ => (true, src),
+        b'-' if is_signed_ty => {
+            let first = digits.get(0).ok_or_else(|| PIE { kind: InvalidDigit })?;
+            digits = &digits[1..];
+            (
+                false,
+                T::from_u32(0)
+                    - T::from_u32(
+                        (*first as char)
+                            .to_digit(radix)
+                            .ok_or_else(|| PIE { kind: InvalidDigit })?,
+                    ),
+            )
+        }
+        val => (
+            true,
+            T::from_u32((val as char).to_digit(radix).ok_or(PIE { kind: InvalidDigit })?),
+        ),
     };
 
-    let mut result = T::from_u32(0);
-
-    if can_not_overflow::<T>(radix, is_signed_ty, digits) {
+    if mem::size_of::<T>() > 2 {
         // If the len of the str is short compared to the range of the type
         // we are parsing into, then we can be certain that an overflow will not occur.
         // This bound is when `radix.pow(digits.len()) - 1 <= T::MAX` but the condition
-        // above is a faster (conservative) approximation of this.
+        // in `safe_width` is a faster (conservative) approximation of this.
         //
         // Consider radix 16 as it has the highest information density per digit and will thus overflow the earliest:
         // `u8::MAX` is `ff` - any str of len 2 is guaranteed to not overflow.
         // `i8::MAX` is `7f` - only a str of len 1 is guaranteed to not overflow.
+        let safe_width = safe_width::<T>(radix, is_signed_ty);
+        
```

Comment:
suggestion: the `.take` in one spot not coupled with a corresponding `.skip` in the other makes this read a bit strangely to me.  Perhaps the splitting could just be put here, with no need to ever look at the length again later?  As a first thought, something like this, with appropriate updates to the `for` loops?
```suggestion
        let (safe_digits, risky_digits) = if safe_width > digits.len() { (digits, &[]) } else { digits.split_at(safe_width) };
                
```